fiftyone/core/odm/document.py

```diff
@@ -649,63 +657,112 @@ def _save(
 
         ensure_connection()
 
-        if validate:
-            self.validate(clean=clean)
-
         id_field = self._meta["id_field"]
-
-        created = self._created
-        if not created:
-            created = "_id" not in self.to_mongo(fields=[id_field])
-
-        doc = self.to_mongo()
-        _id = doc.get("_id", None)
+        _id = self.to_mongo(fields=[id_field]).get("_id", None)
+        created = self._created or _id is None
 
         ops = None
 
         if created:
-            # Save new document
+            # Insert new document
+            if validate:
+                self._validate(clean=clean)
+
+            doc = self.to_mongo()
+
             if _id is None:
                 _id = ObjectId()
                 doc["_id"] = _id
 
-            if deferred:
-                ops = [InsertOne(doc)]
-            else:
-                self._get_collection().insert_one(doc)
+            self[id_field] = self._fields[id_field].to_python(_id)
+            self._created = False
+
+            ops = self._insert(doc, deferred=deferred)
         else:
             # Update existing document
             updates = {}
-            sets, unsets = self._delta()
 
-            if sets:
-                updates["$set"] = sets
+            try:
+                # OPTIMIZATION: we monkey patch `to_mongo()` and
+                # `_get_changed_fields()` here so that mongoengine's
+                # implementations of `_delta()` and `_clear_changed_fields()`,
+                # which call these methods, will not serialize unnecessary
+                # fields and recompute changed fields unecessarily
+                self._get_changed_fields_orig = self._get_changed_fields
+                self._to_mongo_orig = self.to_mongo
+
+                changed_fields = self._get_changed_fields()
+                roots, paths = self._parse_changed_fields(changed_fields)
+
+                if validate and changed_fields:
+                    self._validate_updates(
+                        roots,
+                        paths,
+                        clean=clean,
+                        enforce_read_only=enforce_read_only,
+                    )
+
+                # OPTIMIZATION: apply monkey patch
+                doc = self.to_mongo(fields=roots)
```

Comment:
See https://github.com/voxel51/fiftyone/pull/4729 for full explanation of this super weird behavior. But even though that PR fixes the actual FO usage where this might rear its head, I'm proposing we be doubly safe and follow along with their bad code :( 
Making sure that the doc wouldn't be pruned in this case where we're not fully tracking changes.

```suggestion
                roots = roots if hasattr(self, "_changed_fields") else []
                doc = self.to_mongo(fields=roots)
```

Reply:
I'm still not fully understanding the monoengine weirdness here. If `roots` might not contain all fields, then `_validate_updates()` is not going to successfully validate that other fields contain proper values, right?

Do I need to just rip out this optimization?

Reply:
At minimum, happy to refactor the optimization into a separate PR for consideration.

Reply:
ah yes we would want the altered `roots` to be passed there too.

Basically the weirdness can be summarized by: "if doc does not have `_changed_fields` attribute then it is interpreted the same as `[]` meaning all fields, no matter what `doc._get_changed_fields()` returns"

I think putting into a separate PR for discussion there would be ideal 👍🏼

Reply:
Okay I updated this to handle docs with no `_changed_fields` property (in which case all fields are validated, just to be safe).

I could still move the monkey patch optimization to another PR, but the no `_changed_fields` property code still needs to be here, so the monkey patch is okay to ship IMO.

Reply:
ok